File: Data Structures & Algorithms/reorder-linked-list/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # MUST BE DONE IN PLACE

        # using slow & fast ptr approach, split LL in the middle
        slow, fast = head, head
        while (fast and fast.next):
            slow = slow.next
            fast = fast.next.next

        # reverse linked list w/ prev = None, and curr = slow from step above
        prev, curr = None, slow
        while (curr):
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp

        # connect LL in interweaving fashion
        left, right = head, prev
        while (left != right):
            temp_left = left.next
            if right: temp_right = right.next

            left.next = right
            if right:
                right.next = temp_left

            left = temp_left
            right = temp_right

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode dummy_head(0);
        ListNode* dummy_ptr = &dummy_head;

        while (list1 && list2){
            if (list1->val < list2->val){
                dummy_ptr->next = list1;
                list1 = list1->next;
            } else {                
                dummy_ptr->next = list2;
                list2 = list2->next;
            }
            dummy_ptr = dummy_ptr->next;
        }

        if (list1){
            dummy_ptr->next = list1;
        } else {
            dummy_ptr->next = list2;
        }

        return dummy_head.next;
    }
};
